frontend/src/components/elements/PlotlyChart/CustomTheme.tsx

```diff
@@ -0,0 +1,640 @@
+  const customDataToDataIdx = new Map<string, number[]>()
+  const graphIdxToCustomData = new Map<number, Map<string, number[]>>()
+  data.forEach((graph: any, graphIndex: number) => {
+    if (
+      graph.customdata !== undefined &&
+      graph.marker !== undefined &&
+      Array.isArray(graph.marker.colors) &&
+      graph.marker.colors.length > 0 &&
+      typeof graph.marker.colors[0] !== "number"
+    ) {
+      graph.customdata.forEach((data: any, dataIndex: any) => {
+        const dataString = data.toString()
+        if (Array.isArray(data) && data.length > 0) {
+          if (customDataToDataIdx.has(dataString)) {
+            customDataToDataIdx.set(
+              dataString,
+              // @ts-ignore
+              customDataToDataIdx.get(dataString)?.concat(dataIndex)
+            )
+          } else {
+            customDataToDataIdx.set(dataString, [dataIndex])
+          }
+        }
+      })
+      graphIdxToCustomData.set(graphIndex, customDataToDataIdx)
+    }
+    if (graph.legendgroup !== undefined) {
+      if (legendGroupToIndexes.has(graph.legendgroup)) {
+        legendGroupToIndexes.set(
+          graph.legendgroup,
+          // @ts-ignore
+          legendGroupToIndexes.get(graph.legendgroup).concat(graphIndex)
+        )
+      } else {
+        legendGroupToIndexes.set(graph.legendgroup, [graphIndex])
+      }
+    }
+  })
+
+  let colorIndex = 0
+  legendGroupToIndexes.forEach((dataIdx: number[]) => {
+    dataIdx.forEach((index: number) => {
+      if (data[index].line !== undefined) {
+        // dont assign colors for dist plot boxes when they're transparent
+        if (
+          data[index].type !== "box" &&
+          data[index].line.color !== "rgba(255,255,255,0)" &&
+          data[index].line.color !== "transparent"
+        ) {
+          data[index].line = assign(data[index].line, {
+            color: categoryColors[colorIndex % categoryColors.length],
+          })
+        }
+      }
+      if (
+        data[index].marker !== undefined &&
+        typeof data[index].marker.color === "string"
+      ) {
+        data[index].marker = assign(data[index].marker, {
+          color: categoryColors[colorIndex % categoryColors.length],
+        })
+      }
+    })
+    colorIndex++
+  })
+
+  colorIndex = 0
+  graphIdxToCustomData.forEach(
+    (customData: Map<string, number[]>, dataIndex: number) => {
+      customData.forEach(markerIndexes => {
+        markerIndexes.forEach(markerIndex => {
+          data[dataIndex].marker.colors[markerIndex] =
+            categoryColors[colorIndex % categoryColors.length]
+        })
+        colorIndex++
+      })
+    }
+  )
+}
+
+/**
+ * This overrides the colorscale (continuous colorscale) to all graphs.
+ * @param data - spec.data
+ */
+export function applyColorscale(data: any): any {
+  const theme = useTheme()
+  data.forEach((entry: any) => {
+    entry = assign(entry, {
+      colorscale: hasLightBackgroundColor(theme)
```

Comment:
nit: just to confirm: the `colorscale` is also sometimes directly set in the chart spec without the user having done this explicitly?